### PR TITLE
Fix browser support

### PR DESCRIPTION
**Pull request by mattcollier**
@gannan08 what is the disposition of this code?  Does it need to be cleaned-up in some way.  

What are we doing about the browser context issues?  We're going backwards by adding context docs back into modules.

### CODE REVIEW

**Comment by gannan08** [comment]
> @gannan08 what is the disposition of this code? Does it need to be cleaned-up in some way.

Need to add http-signature-middleware for httpSig support on both the browser and node environment.

> What are we doing about the browser context issues? We're going backwards by adding context docs back into modules.

We want to add them back into modules. Waiting on someone to update those libraries so that they are not using `fs`.

**Comment by mattcollier** [comment]
@gannan08 can the http-signature support be part of another PR?  So we can merge this, and add an issue tracking the context issue?

**Comment by gannan08** [comment]
> @gannan08 can the http-signature support be part of another PR? So we can merge this, and add an issue tracking the context issue?

The http-signature support needs to go into this PR to complete browser support. Tracking the context issue is fine, but I'm unsure as to whether or not we want to make a release with our own hardcoded contexts.

File: README.md [hunk 1]
```diff
@@ -1,37 +1,45 @@
...
 
-For documentation on configuration, see [config.js](./lib/config.js).
+* options - a set of options used when generating the DID Document
+  * didType - the type of DID to generate.
+      Options: 'nym' (default) or 'uuid'
+  * `invokeKey` - optionally pass in a Capability Invocation key, otherwise
+    it will be generated.
+  * keyType - the type of keys to generate.
+      Options: 'Ed25519VerificationKey2018' (default) or
+      'RsaVerificationKey2018'
```
Comment (dlongley) [inline README.md]: I recommend we remove `RsaVerificationKey2018` from the generation options documentation. It's ok to keep the support in there as hidden for now, but I suspect we may not allow RSA based `did:v1:nym` DIDs. That doesn't preclude one from adding the keys after the DID is generated, but RSA-based nyms are messier for a lot of reasons and don't come with the same feature set. Also, ed25519 is slated to get NIST approval in the next fips publication -- should that matter.

**Comment by dlongley** [comment]
Grr, ignore that approval. Apparently github was only showing me a very small subset of the changes here when I approved.

File: README.md (hunk 1 above)
Comment (dmitrizagidulin) [reply README.md]: Removed.

File: lib/VeresOneClient.js [hunk 2]
```diff
@@ -192,26 +189,24 @@ class VeresOneClient {
...
-        headers[header] = value;
-      },
-      method,
-      path
-    }, {
-      headers: signHeaders,
-      keyId,
-      key
+  static async signRequestHeaders({requestOptions, signer}) {
+    if(!requestOptions.headers.Expires) {
```
Comment (dlongley) [inline lib/VeresOneClient.js]:
```suggestion
    // TODO: update to use `(expires)` pseudo header or just use `http-signature-zcap-invoke`
    // ...if possible
    if(!requestOptions.headers.Expires) {
```

File: README.md [hunk 3]
```diff
@@ -1,37 +1,44 @@
...
+// See Configuration below for list of options
+const options = {mode: 'dev', httpsAgent, hostname: 'localhost:12345'};
+const veresDriver = v1.driver(options);
 ```
 
 ## Configuration
 
-For documentation on configuration, see [config.js](./lib/config.js).
+* options - a set of options used when generating the DID Document
+  * didType - the type of DID to generate.
```
Comment (davidlehn) [inline README.md]: Be consistent and use backticks on all these rather than just invokeKey.

File: README.md [hunk 4]
```diff
@@ -41,23 +48,38 @@ resolved Promise.
 
 ### Generate a DID Document
 
-Generate a new DID document.
+#### Generating and Registering a Veres One DID Document
```
Comment (davidlehn) [inline README.md]: The registration part is duplicated in next section.  Can this section be just the generation part?

File: README.md [hunk 5]
```diff
@@ -41,23 +48,38 @@ resolved Promise.
...
-* publicDidDocument - the generated DID Document.
-* privateDidDocument - the DID Document augmented with the encrypted private
-    keys in PEM format.
+To register a DID Document:
+
+```js
+const registrationResult = await veresDriver.register({didDocument});
+```
+
+#### `get()` - Retrieving a Veres One DID Document
```
Comment (davidlehn) [inline README.md]: I don't think the "get()" part needs to be in the section title.  Other sections don't do that.

File: README.md [hunk 6]
```diff
@@ -41,23 +48,38 @@ resolved Promise.
...
+
+```js
+const did = 'did:v1:test:nym:ApvL3PKAzQvFnRVqyZKhSYD2i8XcsLG1Dy4FrSdEKAdR';
+const didDoc = await veresDriver.get({did});
+console.log(JSON.stringify(didDoc, null, 2));
+```
 
 ### Attach an ocap-ld delegation proof to a capability DID Document
```
Comment (davidlehn) [inline README.md]:
```suggestion
### Attach an OCAP-LD delegation proof to a capability DID Document
```

File: README.md (hunk 4 above)
Comment (dmitrizagidulin) [reply README.md]: Fixed, thanks.

File: README.md (hunk 5 above)
Comment (dmitrizagidulin) [reply README.md]: Thanks.

File: README.md (hunk 6 above)
Comment (dmitrizagidulin) [reply README.md]: Fixed, thanks.

File: README.md (hunk 3 above)
Comment (dmitrizagidulin) [reply README.md]: Done, thanks.

**Comment by aljones15** [comment]
wow this makes me feel horrible I never finished that http-signatures using node forge changes. Were sufficient http signature changes made by @dlongley to allow this to merge?

**Comment by dlongley** [comment]
@aljones15 -- not sure to what you're referring, but http-signatures works in the browser just fine now.

File: CHANGELOG.md [hunk 7]
```diff
@@ -16,6 +16,7 @@ browser compatible release will be made soon.
 ## 4.0.0 - 2019-01-30
 
 ### Changed
+- **BREAKING**: Uses `v1.driver(options)` API instead of `v1.veres()`.
```
Comment (mattcollier) [inline CHANGELOG.md]: Is this supposed to be under an upcoming release heading?

File: lib/documentLoader.js [hunk 8]
```diff
@@ -10,10 +10,22 @@ const webLedgerContext = require('web-ledger-context');
 const didContext = require('did-context');
 
 const contextDocuments = new Map([
-  [constants.DID_CONTEXT_URL, didContext],
-  [constants.JSON_LD_PATCH_CONTEXT_V1_URL, jsonldPatchContext],
-  [constants.VERES_ONE_CONTEXT_URL, veresOneContext],
-  [constants.WEB_LEDGER_CONTEXT_URL, webLedgerContext],
+  [
```
Comment (mattcollier) [inline lib/documentLoader.js]: An optimization for all these contexts that would reduce the boiler plate.  All the `foo.contexts` are maps.
```
[
  ...didContext.contexts,
  ...jsonldPatchContext.contexts,
  ...veresOneContext.contexts,
  ...webLedgerContext.contexts
]
```

File: CHANGELOG.md (hunk 7 above)
Comment (dmitrizagidulin) [reply CHANGELOG.md]: Oh, hmm, that should probably be changed to v6, yeah?

Comment (mattcollier) [reply CHANGELOG.md]: 6.0 is already released, this would be 7.0.  Which makes me wonder if this whole branch needs to be rebased?

Comment (dmitrizagidulin) [reply CHANGELOG.md]: I just rebased this branch on the latest master yesterday, so it should be fine.
(And yeah, I'll change it to v7. I was confused looking at the 6.0.1 thing)

File: lib/documentLoader.js (hunk 8 above)
Comment (mattcollier) [reply lib/documentLoader.js]: Another nice side-effect to this method is that we pick up contexts for modules that are exporting multiple contexts automatically.  The security context for example.

Comment (dmitrizagidulin) [reply lib/documentLoader.js]: Changed, thanks

File: CHANGELOG.md (hunk 7 above)
Comment (dmitrizagidulin) [reply CHANGELOG.md]: Fixed.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,10 @@
 # did-veres-one ChangeLog
 
+## 7.0.0 - 2019-08-01
```
Comment (mattcollier) [inline CHANGELOG.md]: Leave date "TBD" or somesuch.  For the release flow, it is the hope that the person doing the release *just* has to set this date properly and then run the release script.  The changelog *must* have unstaged changes for the release script to work.

Comment (dmitrizagidulin) [reply CHANGELOG.md]: Np, changed to TBD.

**Comment by dmitrizagidulin** [comment]
@mattcollier 
> My name is mattcollier and I approve these changes.

lol, excellent. (Yeah, it's confusing, they put the button to add yourself as the reviewer only on the Files tab.)

**Pull request merged**